examples/cmd/logtransport.go

```diff
@@ -22,15 +21,23 @@ type logTransport struct {
 func (t *logTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	t0 := time.Now()
 
-	reqBody, _ := ioutil.ReadAll(r.Body)
-	r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
-
+	// Since the underlaying Transport.RoundTrip might consume the Request, and it's also needed later to log the process,
```

Comment:
I don't understand what you do here.

Logger should log the request it received and shouldn't care about what is going on at the level below. If you need to log after the modifications by other transports just put the logger transport deeper.

Reply:
The `logTransport` is already the deepest one.
The next level, is the `http.client.Transport.RoundTrip`, which will call go code from the stdlib.

The problem is that when the request is processed is sent, it is read, so it will be no longer readable after it is sent wich affects us in two ways:
- it can not be logged after it is sent,
- it can not be accessed later by any other Transports at a higher level (this is the deepest one)

You can reproduce it with this [snippet command](https://gist.github.com/dpordomingo/0f1ef5d371e2862e47e59aad56be8134) or use it as a shorter, and documented, example of the problem. If you change to `false` any of `backupRequestAfterLeavingLastTransport`, `backupRequestAfterReadingIt` or `backupResponseAfterReadingIt` you'll see that the process will fail at some point.

Reply:
I still don't understand you. The logger must do only 1 job as I described above. Log request and response without mumbo-jumbo with the request after it was processed by an underlying round-tripper.
The original implementation is absolutely correct.

if there is any other problem the fix must be done where the problem is introduced not in the logger. As I don't really understand you, I can't point where exactly. But most probably in the retry transport.

Reply:
I also don't understand, since you're wrapping with a `NopCloser` it should still be readable. So doing like:
```
	reqBody, _ := ioutil.ReadAll(r.Body)
	r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))

	resp, err := t.T.RoundTrip(r)
```
shouldn't ruin the request body.

Reply:
done following the alternative you suggested 2717914